File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        /*int n = nums.length;
        int[] result = new int[n];
        for(int i = 0;i<n;i++){
            int product = 1;
        for(int j= 0;j<n;j++){
            if(i != j){
                product = product*nums[j]; //BRUTE FORCE APPROACH 
            }                                TIME COMPLEXITY:0(n^2)
        }                                     SPACE COMPLEXITY:0(n)
        result[i] = product;
        }
    return result;*/



//  OPTIMIZED APPROACH
        int n = nums.length;

        int[] result = new int[n];

        // STEP 1: Store left products
        result[0] = 1;

        for(int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];//TIME COMPLEXITY:0(n)
        }                                           //SPACE COMPLEXITY:0(1)

        // STEP 2: Multiply right products
        int rightProduct = 1;

        for(int i = n - 1; i >= 0; i--) {

            result[i] = result[i] * rightProduct;

            rightProduct = rightProduct * nums[i];
        }

        return result;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
'''class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)

        prefix = [1] * n
        suffix = [1] * n
        answer = [1] * n

        # Build prefix array
        product = 1

        for i in range(n):
            prefix[i] = product
            product = product * nums[i]

        # Build suffix array
        product = 1

        for i in range(n - 1, -1, -1):
            suffix[i] = product
            product = product * nums[i]

        # Build answer
        for i in range(n):
            answer[i] = prefix[i] * suffix[i]

        return answer'''
class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)

        answer = [1] * n

        # Prefix products
        product = 1

        for i in range(n):
            answer[i] = product
            product = product * nums[i]

        # Suffix products
        right_product = 1

        for i in range(n - 1, -1, -1):
            answer[i] = answer[i] * right_product
            right_product = right_product * nums[i]

        return answer